service/frontend/workflowHandler.go

```diff
@@ -1620,6 +1623,166 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
+			// we guarantee to use the first event ID provided in the request
+			request.GetFirstEventId() != token.FirstEventID ||
+			// the next event ID in the request must be <= next event ID from mutable state, when initialized
+			// so as long as customer do not change next event ID during pagination,
+			// next event ID in the token <= next event ID in the request.
+			request.GetNextEventId() < token.NextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid pagination token."}
+		}
+
+		// for the rest variables in the token, since we do not do hmac,
+		// the only thing can be done is to trust the token:
+		// IsWorkflowRunning: not used
+		// TransientDecision: not used
+		// PersistenceToken: trust
+		// EventStoreVersion: trust
+		// ReplicationInfo: trust
+
+	} else {
+		firstEventID := request.GetFirstEventId()
+		nextEventID := request.GetNextEventId()
+		if firstEventID < 0 || firstEventID > nextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid FirstEventID && NextEventID combination."}
+		}
+
+		response, err := wh.history.GetMutableState(ctx, &h.GetMutableStateRequest{
+			DomainUUID:          common.StringPtr(domainID),
+			Execution:           execution,
+			ExpectedNextEventId: common.Int64Ptr(common.FirstEventID), // common.FirstEventID means no long poll
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// branch token is provided as part of request, makeing sure that no branching has happened yet
+		// otherwise, caller need to do a reload of history from the begining
+		if request.BranchToken != nil && bytes.Compare(request.BranchToken, response.BranchToken) != 0 {
+			return nil, &gen.BadRequestError{Message: "History has encounter branching event, should reload history from begining."}
+		}
+
+		// check if the input next event ID is > actual next event ID in the mutable state
+		// since we should not leak invalid events
+		if nextEventID > response.GetNextEventId() {
+			nextEventID = response.GetNextEventId()
+		}
+		token = &getHistoryContinuationToken{
+			RunID:             execution.GetRunId(),
+			BranchToken:       response.BranchToken,
+			FirstEventID:      firstEventID,
+			NextEventID:       nextEventID,
+			PersistenceToken:  nil, // this is the initialized value
+			EventStoreVersion: response.GetEventStoreVersion(),
+			ReplicationInfo:   response.ReplicationInfo,
+		}
+	}
+
+	if token.FirstEventID >= token.NextEventID {
+		return &gen.GetWorkflowExecutionRawHistoryResponse{
+			BranchToken:       token.BranchToken,
+			HistoryBatches:    []*gen.DataBlob{},
+			ReplicationInfo:   token.ReplicationInfo,
+			EventStoreVersion: common.Int32Ptr(token.EventStoreVersion),
+			NextPageToken:     nil, // no further pagination
+		}, nil
+	}
+
+	// TODO need to deal with transient decision if to be used by client getting history
+	var historyBatches []*gen.History
+	_, historyBatches, token.PersistenceToken, _, err = historyService.PaginateHistory(
```

Comment:
This defeats the purpose of RawHistory API.  If we are already deserializing history before sending it back to caller then I don't see a point in exposing this GetRawHistory API.

Reply:
the refactoring of the persistent layer providing raw & valid data is an separate task, and there is already an issue for it: https://github.com/uber/cadence/issues/1323

for your question, this does not defeat the purpose of sending raw history, since worker in most cases, do not need to do serialization & deserialization.

moreover, our persistent layer does not guarantee returning the correct raw data now, since there is fair amount of logic filtering out the stale data, plus, event store v1 and v2 do not have any unit test.